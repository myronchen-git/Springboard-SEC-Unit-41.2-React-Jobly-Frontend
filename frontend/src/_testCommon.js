const companies = Object.freeze([
  Object.freeze({
    handle: 'comp1',
    name: 'Company 1',
    description: 'Company 1 description.',
    numEmployees: 12,
    logoUrl: 'logo1.png',
  }),
  Object.freeze({
    handle: 'comp2',
    name: 'Company 2',
    description: 'Company 2 description.',
    numEmployees: 123,
    logoUrl: 'logo2.png',
  }),
  Object.freeze({
    handle: 'comp3',
    name: 'Company 3',
    description: 'Company 3 description.',
    numEmployees: 3,
  }),
  Object.freeze({
    handle: 'comp4',
    name: 'Company 4',
    description: 'Company 4 description.',
    logoUrl: 'logo4.png',
  }),
]);

const jobs = Object.freeze([
  Object.freeze({
    id: '1',
    title: 'Job 1',
    salary: 1,
    equity: 0.1,
    companyHandle: 'comp1',
    companyName: 'Company 1',
  }),
  Object.freeze({
    id: '2',
    title: 'Job 2',
    salary: 2,
    equity: 0.2,
    companyHandle: 'comp1',
    companyName: 'Company 1',
  }),
]);

const comp1Jobs = [];
[jobs[0], jobs[1]].forEach((job) => {
  const condensedJob = { ...job };
  delete condensedJob.companyHandle;
  delete condensedJob.companyName;
  comp1Jobs.push(Object.freeze(condensedJob));
});

const companyDetails = Object.freeze({
  handle: 'comp1',
  name: 'Company 1',
  description: 'Company 1 description.',
  numEmployees: 12,
  logoUrl: 'logo1.png',
  jobs: Object.freeze(comp1Jobs),
});

// ==================================================

export { companies, companyDetails, jobs };
