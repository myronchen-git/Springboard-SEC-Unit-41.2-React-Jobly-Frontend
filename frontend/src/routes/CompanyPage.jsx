import PropTypes from 'prop-types';
import { useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';

import JoblyApi from '../../../api';

// ==================================================

/**
 * Displays detailed info about a company and a list of its job openings.
 *
 * @param {Object} props - React component properties.
 * @param {Array} props.applications - A list of job IDs that were applied to.
 */
function CompanyPage({ applications }) {
  const { handle } = useParams();
  const [company, setCompany] = useState({});
  const [jobs, setJobs] = useState([]);

  useEffect(() => {
    JoblyApi.getCompany(handle).then((companyData) => {
      setJobs([...companyData.jobs]);
      delete companyData.jobs;
      setCompany(companyData);
    });
  }, [handle]);

  return (
    <main className="CompanyPage">
      <section className="CompanyPage__info">
        <header>
          <img src={company?.logoUrl} alt={`${company.name} Logo`} />
          <h1>{company.name}</h1>
        </header>
        {company.numEmployees && <p>{company.numEmployees} employees</p>}
        <p>{company.description}</p>
      </section>
      <ul className="CompanyPage__jobs">
        {jobs.map((job) => (
          <li key={job.id}>
            <p>{job.title}</p>
          </li>
        ))}
      </ul>
    </main>
  );
}

// ==================================================

CompanyPage.propTypes = {
  applications: PropTypes.array.isRequired,
};

export default CompanyPage;
