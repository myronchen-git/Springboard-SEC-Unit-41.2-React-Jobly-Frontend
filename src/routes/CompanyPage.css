.CompanyPage__info header {
  margin: 1em auto;
}

.CompanyPage__logo-container {
  margin: 1em;
}

.CompanyPage__logo {
  height: 100px;
  width: 100px;
}
