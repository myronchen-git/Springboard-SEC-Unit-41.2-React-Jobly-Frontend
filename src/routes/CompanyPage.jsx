import { useContext, useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';

import JoblyApi from '../api.js';
import JobCard from '../components/JobCard.jsx';
import { UserContext } from '../contexts.jsx';

// ==================================================

/**
 * Displays detailed info about a company and a list of its job openings.
 */
function CompanyPage() {
  const { handle } = useParams();
  const [company, setCompany] = useState({});
  const [jobs, setJobs] = useState([]);
  const { applications } = useContext(UserContext);
  const [errorMessage, setErrorMessage] = useState();

  useEffect(() => {
    JoblyApi.getCompany(handle)
      .then((companyData) => {
        setJobs([...companyData.jobs]);
        delete companyData.jobs;
        setCompany(companyData);
      })
      .catch((err) => setErrorMessage(err[0]));
  }, [handle]);

  return (
    <main className="CompanyPage">
      {errorMessage ? (
        <div className="CompanyPage__error">
          <p>{errorMessage}</p>
        </div>
      ) : (
        <>
          <section className="CompanyPage__info">
            <header>
              <img src={company?.logoUrl} alt={`${company.name} Logo`} />
              <h1>{company.name}</h1>
            </header>
            {company.numEmployees && <p>{company.numEmployees} employees</p>}
            <p>{company.description}</p>
          </section>
          <ul className="CompanyPage__jobs">
            {jobs.map((job) => (
              <li key={job.id}>
                <JobCard job={job} isApplied={applications.includes(job.id)} />
              </li>
            ))}
          </ul>
        </>
      )}
    </main>
  );
}

// ==================================================

export default CompanyPage;
