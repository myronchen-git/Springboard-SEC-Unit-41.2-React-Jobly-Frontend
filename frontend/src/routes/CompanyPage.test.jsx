import { act, render } from '@testing-library/react';
import { useParams } from 'react-router-dom';
import { beforeEach, describe, expect, it, vi } from 'vitest';

import JoblyApi from '../../../api';
import { companyDetails } from '../_testCommon.js';
import CompanyPage from './CompanyPage.jsx';

// ==================================================

vi.mock('react-router-dom');

vi.mock(import('../../../api'), () => {
  const MockJoblyApi = vi.fn();
  MockJoblyApi.getCompany = vi.fn();

  return {
    default: MockJoblyApi,
  };
});

// ==================================================

describe('CompanyPage', () => {
  const applications = Object.freeze([2, 3]);

  beforeEach(() => {
    useParams.mockReturnValue({ handle: companyDetails.handle });
    JoblyApi.getCompany.mockResolvedValue(structuredClone(companyDetails));
  });

  it('renders.', async () => {
    await act(async () => render(<CompanyPage applications={applications} />));
  });

  it('matches snapshot.', async () => {
    const { asFragment } = await act(async () =>
      render(<CompanyPage applications={applications} />)
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays data about company.', async () => {
    // Act
    const { getByAltText, getByText } = await act(async () =>
      render(<CompanyPage applications={applications} />)
    );

    // Assert
    expect(getByAltText(`${companyDetails.name} Logo`)).toBeVisible();
    expect(getByText(companyDetails.name)).toBeVisible();
    expect(getByText(companyDetails.description)).toBeVisible();
    expect(getByText(companyDetails.numEmployees + ' employees')).toBeVisible();

    companyDetails.jobs.forEach((job) => {
      expect(getByText(job.title)).toBeVisible();
    });
  });
});
