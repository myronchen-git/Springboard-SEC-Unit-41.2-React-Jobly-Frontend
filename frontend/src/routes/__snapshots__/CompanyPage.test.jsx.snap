// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CompanyPage > matches snapshot. 1`] = `
<DocumentFragment>
  <main
    class="CompanyPage"
  >
    <section
      class="CompanyPage__info"
    >
      <header>
        <img
          alt="Company 1 Logo"
          src="logo1.png"
        />
        <h1>
          Company 1
        </h1>
      </header>
      <p>
        12 employees
      </p>
      <p>
        Company 1 description.
      </p>
    </section>
    <ul
      class="CompanyPage__jobs"
    >
      <li>
        <article
          class="JobCard"
        >
          <header
            class="JobCard__header"
          >
            <h1>
              Job 1
            </h1>
          </header>
          <summary
            class="JobCard__info"
          >
            <p>
              Salary: 1
            </p>
            <p>
              Equity: 0.1
            </p>
          </summary>
          bound HTMLFormElement {
            "0": <button
              type="button"
            >
              Apply
            </button>,
          }
        </article>
      </li>
      <li>
        <article
          class="JobCard"
        >
          <header
            class="JobCard__header"
          >
            <h1>
              Job 2
            </h1>
          </header>
          <summary
            class="JobCard__info"
          >
            <p>
              Salary: 2
            </p>
            <p>
              Equity: 0.2
            </p>
          </summary>
          bound HTMLFormElement {
            "0": <button
              disabled=""
              type="button"
            >
              Applied
            </button>,
          }
        </article>
      </li>
    </ul>
  </main>
</DocumentFragment>
`;
